src/drivers/fpga/socfpga/socfpga.c

```diff
@@ -0,0 +1,355 @@
+	log_debug(" GPIO_PORTA_EOI   %08x", socfpga_read(socfpga, SOCFPGA_MGR_GPIO_PORTA_EOI));
+	log_debug(" GPIO_EXT_PORTA   %08x", socfpga_read(socfpga, SOCFPGA_MGR_GPIO_EXT_PORTA));
+	log_debug(" GPIO_LS_SYNC     %08x", socfpga_read(socfpga, SOCFPGA_MGR_GPIO_LS_SYNC));
+	log_debug(" GPIO_VER_ID_CODE %08x", socfpga_read(socfpga, SOCFPGA_MGR_GPIO_VER_ID_CODE));
+	log_debug(" GPIO_CONFIG_REG2 %08x", socfpga_read(socfpga, SOCFPGA_MGR_GPIO_CONFIG_REG2));
+	log_debug(" GPIO_CONFIG_REG1 %08x", socfpga_read(socfpga, SOCFPGA_MGR_GPIO_CONFIG_REG1));
+}
+
+static uint32_t socfpga_mode(struct socfpga *socfpga) {
+	uint32_t ret = socfpga_read(socfpga,
+			SOCFPGA_MGR_STAT & SOCFPGA_MODE_MASK);
+	switch (ret) {
+	case SOCFPGA_MODE_MASK:
+	case SOCFPGA_MODE_OFF:
+	case SOCFPGA_MODE_RESET:
+	case SOCFPGA_MODE_CONFIG:
+	case SOCFPGA_MODE_INIT:
+	case SOCFPGA_MODE_USER:
+	case SOCFPGA_MODE_UNKNOWN:
+		return ret;
+	default:
+		log_error("incorrect socfpga state: 0x%x", ret);
+		return 0;
+	}
+}
+
+static int socfpga_mode_wait(struct socfpga *socfpga, uint32_t mode) {
+	int t = TIMEOUT;
+
+	while (t--) {
+		if (socfpga_mode(socfpga) == mode) {
+			break;
+		}
+
+		usleep(100);
+	}
+
+	if (t <= 0) {
+		log_error("FPGA reset timeout!");
+		return -ETIMEDOUT;
+	}
+
+	return 0;
+}
+
+static int socfpga_reset(struct socfpga *socfpga) {
+	/* Enable FPGA.
+	 * Note: this step is not mentioned in datasheet,
+	 * but is done in Linux driver */
+	socfpga_or(socfpga, SOCFPGA_MGR_CTRL,
+			SOCFPGA_CTRL_EN);
+
+	/* Enter reset state */
+	socfpga_or(socfpga, SOCFPGA_MGR_CTRL,
+			SOCFPGA_CTRL_NCONFIGPULL);
+
+	if (socfpga_mode_wait(socfpga, SOCFPGA_MODE_RESET)) {
+		return -ETIMEDOUT;
+	}
+
+	/* Exit reset phase */
+	socfpga_clear(socfpga, SOCFPGA_MGR_CTRL,
+			SOCFPGA_CTRL_NCONFIGPULL);
+
+	return 0;
+}
+
+
+/**
+ * @brief Corresponds to Power Up Phase, Reset Phase in docs
+ */
+static int socfpga_config_init(struct fpga *fpga) {
+	struct socfpga *socfpga = fpga->priv;
+	uint32_t tmp;
+	int err;
+
+	/* TODO: figure out what values should be actualy passed.
+	 * According to manual, "You must set the cdratio and
+	 * cfgwdth bits appropriatly" */
+	tmp = socfpga_read(socfpga, SOCFPGA_MGR_CTRL);
+	tmp &= ~SOCFPGA_CTRL_CDRATIO_OFFT;
+	tmp &= ~SOCFPGA_CTRL_CFGWDTH;
+	/* TODO: Set correct values */
+	socfpga_write(socfpga, SOCFPGA_MGR_CTRL, tmp);
+
+	err = socfpga_reset(socfpga);
+	if (err != 0) {
+		return err;
+	}
+
+	err = socfpga_mode_wait(socfpga, SOCFPGA_MODE_CONFIG);
+	if (err != 0) {
+		return -ETIMEDOUT;
+	}
+
+	/* Clear interrupt status */
+
+	/* Enable data transfer to FPGA */
+	socfpga_or(socfpga, SOCFPGA_MGR_CTRL, SOCFPGA_CTRL_AXICFEN);
+
+	return 0;
+}
+
+/**
+ * @brief Corresponds to Configuration Phase in docs
+ */
+static int socfpga_config_write(struct fpga *fpga, const uint8_t *buf, size_t len) {
+	struct socfpga *socfpga = fpga->priv;
+	uint32_t *buf32 = (uint32_t *) buf;
+	uint32_t tmp;
+
+	while (len > sizeof(uint32_t)) {
```

Comment:
may be for will look better?

Reply:
I think that 4-byte alignment check will look uglier with for loop, and it doesn't really make big difference to use `for` or `while` here.